### PR TITLE
 importer: replace Clear/RevertRange calls with DeleteRange in IMPORT rollback

### PR DESCRIPTION
**Pull request by msbutler**
This patch replaces the ClearRange and RevertRange calls with MVCC DeleteRange
during IMPORT rollbacks.

In the predicate based DeleteRange call, the client's provided table span is
partitioned in order to process the partitions in parallel. The new
bulkio.import.predicate_delete_range_parallelism setting defines the number of
workers that can issue DeleteRange requests in concurrently, and the new
bulkio.import.predicate_delete_range_batch_size setting defines the number of
ranges of data to include in a single DeleteRange request.

Informs #70428

Release note: none

### CODE REVIEW

**Comment by msbutler** [comment]
RFAL!

File: pkg/sql/rollback.go
Comment (msbutler) [inline pkg/sql/rollback.go:141]: Currently, I buffer up all the partitioned spans before issuing requests, which could be memory intensive for a big enough table. I tried putting this initial span sending into a separate go routine to reduce the span buffer size and to allow span partition creation and DeleteRange requests to occur simultaneously, but I couldn't think of a good design after an hour.  If folks think this is a big problem, would be great to pair program on this. I don't have much experience working channels.

File: pkg/sql/importer/import_job.go
Comment (erikgrinaker) [inline pkg/sql/importer/import_job.go:1469]: I think it might be prudent to have an internal cluster setting for this too, in case we find something horribly wrong.

@dt Any good reason why we wouldn't? Like, would it be helpful to have a guarantee that all operations are MVCC-compliant by 23.1?

File: pkg/sql/importer/import_stmt_test.go
Comment (erikgrinaker) [inline pkg/sql/importer/import_stmt_test.go:3160]: We should have another pair of subtests that enable/disable use of DeleteRange. Probably easier to implement if we also add that cluster setting.

File: pkg/sql/rollback.go
Comment (erikgrinaker) [inline pkg/sql/rollback.go:34]: Wow, those are some big batches. Glad we have byte limits on the server side.

File: pkg/sql/rollback.go
Comment (erikgrinaker) [reply pkg/sql/rollback.go:141]: We don't really need a separate goroutine, we can use the current one. First create the channel, then spin off the workers (who will wait on the channel), and then just feed the spans to the channels in the for loop (where you currently append to `initialSpans`). When done, close the channel and call `Wait()` on the workers.

File: pkg/sql/rollback.go
Comment (erikgrinaker) [inline pkg/sql/rollback.go:139]: Is 100 a bit much? Maybe 10 or 20? Do we have any experience with similar settings elsewhere?

File: pkg/sql/rollback.go
Comment (erikgrinaker) [inline pkg/sql/rollback.go:191]: Is it worth asserting that the version gate is active here, or earlier in the function? If it wasn't, we'd be deleting all the data in the range.

File: pkg/sql/rollback.go
Comment (erikgrinaker) [inline pkg/sql/rollback.go:216]: Oh, I see why this gets tricky now. Would it be better to use a for loop for this rather than feeding it back into the channel? Having the same goroutine both consume and produce messages on a channel is a bit of an antipattern (it muddies the waters wrt. the termination condition, because it isn't clear whose responsibility it is to close the channel).

File: pkg/sql/importer/import_job.go
Comment (msbutler) [reply pkg/sql/importer/import_job.go:1469]: Seems reasonable to me; i'll defer to @dt here.

File: pkg/sql/importer/import_stmt_test.go
Comment (msbutler) [reply pkg/sql/importer/import_stmt_test.go:3160]: agreed. but will wait to implement until cluster setting question is answered.

File: pkg/sql/rollback.go
Comment (msbutler) [reply pkg/sql/rollback.go:145]: Aha! Thanks for the insight! Code looks much cleaner now.

Aside: I based my initial approach on some examples in the bulk processors, one of which uses the antipattern of a go thread sending and receiving on the same channel. I wonder if that should change 🤔

File: pkg/sql/rollback.go
Comment (msbutler) [reply pkg/sql/rollback.go]: @dt any prior here? If not, i can ask KV, or tune this during stability.

File: pkg/sql/rollback.go
Comment (msbutler) [reply pkg/sql/rollback.go]: hmmmm. The caller already does this here and import rollbacks will be the only caller of this funciton.

https://github.com/cockroachdb/cockroach/blob/eb8d649813bd3ded2f80cba340252d553a28b85d/pkg/sql/importer/import_job.go#L1466-L1469

I suppose we could pass a handler to cluster settings to the function, but perhaps it's overkill?

File: pkg/sql/rollback.go
Comment (msbutler) [reply pkg/sql/rollback.go]: thanks for the wisdom!

File: pkg/sql/rollback.go
Comment (erikgrinaker) [reply pkg/sql/rollback.go]: Not a big deal, as long as we're reasonably certain that callers won't forget to check the gate.

**Review by erikgrinaker — COMMENTED** [review]
Did another pass, this all seems fine except for the setting and goroutine leaks discussed separately. Will approve once everything lands.

File: pkg/sql/rollback.go
Comment (msbutler) [reply pkg/sql/rollback.go:230]: nice catch! Done.

File: pkg/sql/importer/import_stmt_test.go
Comment (msbutler) [reply pkg/sql/importer/import_stmt_test.go:3160]: No cluster setting for now. May add during stability.

File: pkg/sql/importer/import_job.go
Comment (msbutler) [reply pkg/sql/importer/import_job.go:1469]: As discussed in MVCC ops meeting, no cluster setting for now, but may add during stability.

**Review by erikgrinaker — APPROVED** [review]
LGTM, but should get an approval from bulk as well.

File: pkg/sql/rollback.go
Comment (erikgrinaker) [reply pkg/sql/rollback.go]: >Aside: I based my initial approach on some examples in the bulk processors, one of which uses the antipattern of a go thread sending and receiving on the same channel. I wonder if that should change 🤔

Right. It's sometimes warranted, but usually it's a good idea to avoid it.

**Review by dt — APPROVED** [review]

**Comment by msbutler** [comment]
unrelated flake

bors r=dt, erikgrinaker

**Pull request merged**
